File: problems/LeetCode/p19.py
# Copyright (c) 2022 Shudipto Trafder
# 
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT

# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(-1, head)
        slow, fast = dummy, head
        
        while n > 0 and fast:
            fast = fast.next
            n -= 1

        while fast:
            fast = fast.next
            slow = slow.next

        # slow.next is the node, we have to remove
        slow.next = slow.next.next

        return dummy.next


if __name__ == "__main__":
    head = ListNode(1)
    temp = head
    for i in range(2, 10):
        temp.next = ListNode(i)
        temp = temp.next
    
    temp = head

    while temp:
        print(temp.val, sep=",")
        temp = temp.next

    head = Solution().removeNthFromEnd(head, 2)
    
    print("AFTER")
    temp = head
    while temp:
        print(temp.val, sep=",")
        temp = temp.next

File: LeetCode-Java/p19.java
// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; this.next = null; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(-1, head);
        ListNode slow = dummy, fast = head;

        while (n > 0 && fast != null) {
            fast = fast.next;
            n--;
        }

        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        // slow.next is the node we have to remove
        slow.next = slow.next.next;

        return dummy.next;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode temp = head;
        for (int i = 2; i < 10; i++) {
            temp.next = new ListNode(i);
            temp = temp.next;
        }

        temp = head;
        while (temp != null) {
            System.out.print(temp.val + " ");
            temp = temp.next;
        }

        head = new Solution().removeNthFromEnd(head, 2);

        System.out.println("\nAFTER");
        temp = head;
        while (temp != null) {
            System.out.print(temp.val + " ");
            temp = temp.next;
        }
    }
}